src/utils/requestUtils.ts

```diff
@@ -48,4 +50,43 @@ export namespace requestUtils {
             stream.pipe(fse.createWriteStream(filePath).on('finish', resolve).on('error', reject));
         });
     }
+
+    // tslint:disable-next-line:no-any
+    export async function getFunctionAppMasterKey(azureSession: AzureSession, resourceId: string, accessToken: any): Promise<HttpOperationResponse> {
+        const subscriptionId: string = getSubscriptionFromId(resourceId, false);
+        const subscriptionContext: ISubscriptionContext = {
```

Comment:
You're doing a lot of stuff manually here that doesn't need to be done that way. For example, use `ext.tree.findTreeItem` to get a `SlotTreeItemBase` matching `resourceId ` and then use `SlotTreeItemBase.client.listHostKeys`

Reply:
@ejizba I see. In our case we are not in a local project yet, but setting it up. Does the ext.tree.findTreeItem work even before the local project is setup?

Reply:
We basically center our entire extension around the tree, so yes you should be able to get `findTreeItem` to work. In this case you're trying to get their function app node, which is an Azure resource that should be there as long as they're signed in, regardless of anything related to the local project. Could be a problem with your resourceId 🤷‍♂️  I would suggest using the debugger or looking at other examples in the codebase.

Reply:
@ejizba In our case, we are passing the correct resourceId and the context as below based on other examples I saw in the code:
const slotTreeItem: SlotTreeItemBase | undefined = await ext.tree.findTreeItem(resourceId, { ...actionContext, loadAll: true });

However, stepping through and debugging the findTreeItem function seems to get only subscription level cachedChildren on line 206 of F:\repos\vscode-azurefunctions\node_modules\vscode-azureextensionui\out\src\tree\AzExtTreeDataProvider.js
It does not get any resourceId specific cachedChildren and hence always returns undefined.

It almost seems like the tree is not fully loaded in our specific case.

Reply:
Just push up what you have so far and I can try it out

Reply:
I am adding the code to findTreeItem for you to try it out. However, I am keeping my existing code as well for now so it does not break our entire scenario. I can clean it up once findTreeItem is working. Also another thing to note is that this section of code is executed before VSCode reloads and initProjectForVSCode gets called.

In order to trigger our scenario you will need to use the below:
vscode://ms-azuretools.vscode-azurefunctions/?resourceId=<appResourceId>&defaultHostName=<appHostName>&devcontainer=<devContainerName>&language=<appLanguage>&downloadAppContent=<true/false>

Reply:
As suspected, it is a problem with the resourceId. [This line](https://github.com/microsoft/vscode-azurefunctions/pull/2568/files#diff-bec5299b320724002801862064e523e95e27d75971447e824195446c0a658b11R108) is converting it to lowercase, but our tree uses normal casing as given to us by Azure APIs and as seen in the portal (i.e. `resourceGroups` instead of `resourcegroups`). If I keep the casing correct, `slotTreeItem` is defined for me.

Reply:
I can update the code to pass in the resourceId before we convert it. However, based on ARM's standards all resourceId's are treated case insensitive as resourceId's are unique irrespective of casing - https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/resource-name-rules.
It is always good engineering practice to make unique ids case insensitive.

Currently maybe there is no scenario that gets affected by this in VSCode, but, in future if there is a scenario of passing in resourceId from an external client and the resourceId casings are changed on the client side, then VSCode side will break.

I have created an issue for tracking this - https://github.com/microsoft/vscode-azurefunctions/issues/2617